### DISCUSSION TITLE
Quarto blog https://fonts.googleapis.com/, causing the site to open very slowly

### DISCUSSION THREAD

**Discussion by chuxinyuan** [quarto-dev/quarto-cli#9321] [Q&A] [answered]
### Description

Quarto blog https://fonts.googleapis.com/, causing the site to open very slowly

Suggest replace with <https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/all.min.css>

**Comment by mcanouil** [comment]
Do you have any benchmark / example showing the issue?

**Reply by chuxinyuan** [reply]
![截图 2024-04-10 16-14-59](https://github.com/quarto-dev/quarto-cli/assets/26940042/9a381a4a-b97c-4ad8-ab55-d7a296b0745f)
![截图 2024-04-10 16-17-52](https://github.com/quarto-dev/quarto-cli/assets/26940042/e89b6346-947d-45c1-9aa0-620ee214b551)

**Reply by chuxinyuan** [reply]
I am in China now.
Test url: <https://deepin.netlify.app/>

**Reply by mcanouil** [reply]
Note that the timing you show are not Font Awesome. Only Bootstrap are in your screenshot.
And Bootstrap is not loaded from a CDN as it is embedded in the website.

> causing the site to open very slowly

Less than half a second to load the page is "very slow"? (Simply trying to understand what "very slow" means objectively/concretely)

I also don't have the errors from your screenshots beside the missing favicon. (I am in Europe/France)
![image](https://github.com/quarto-dev/quarto-cli/assets/8896044/10af0181-4a18-467b-89c6-59bca4244296)

**Reply by chuxinyuan** [reply]
![1111111](https://github.com/quarto-dev/quarto-cli/assets/26940042/cdbbc2f1-09a6-4cbd-b0d3-2131c81f21e5)
RStudo, show in new window.
Maybe I should have used this picture from the beginning.

**Reply by mcanouil** [reply]
I don't think serving the website through localhost and RStudio can really be used as a reference benchmark, as you add several other components which might interact leading to the observed "slow loading".
Also, it seems weird that there are that amounts of queries for the font.

**Reply by chuxinyuan** [reply]
There's no need to knock if the door is open. In fact, I told you that I was in China for a reason, and there was no way to use any of Google's services in China.

**Reply by mcanouil** [reply]
> I told you that I was in China for a reason, and there was no way to use any of Google's services in China.

I have absolutely no idea of what works or not in China and do not believe this is common knowledge.
What seems obvious to you is not necessarily obvious to everyone else.

**Comment by cderv** [comment, marked as answer]
the `fonts.googleapis.com` is the default for web font used by Bootswatch themes. It is quite advanced, but you can customized anything at the SASS level for bootstrap and Bootswatch. We explain this at https://quarto.org/docs/output-formats/html-themes.html#theme-options https://quarto.org/docs/output-formats/html-themes.html#custom-themes and https://quarto.org/docs/output-formats/html-themes-more.html 

Looking as source scss file for bootwatch theme, it seems you can configure the webfont URL path, or even not configure one (to use local font). 

https://github.com/quarto-dev/quarto-cli/blob/6f233b161d770a76db9b5b3a7755fd8f88641ff3/src/resources/formats/html/bootstrap/themes/sandstone.scss#L143-L146

So you can definitely change that for your theme to not use this URL. You would set the variable to another URL with the font available. 
Not that `Source Sans Pro` is the default font also 
https://github.com/quarto-dev/quarto-cli/blob/284a82efa03d5003a09335118b8017312fc97790/src/resources/formats/html/bootstrap/themes/cosmo.scss#L54

but you can change the font easily too 
https://quarto.org/docs/output-formats/html-themes.html#fonts

Hope this helps

**Reply by chuxinyuan** [reply]
<https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/all.min.css> or would it be better to keep the font locally?

**Reply by cderv** [reply]
Your link is about fontawesome fonts. This is not the same at all. 

If you want to use fontawesome, you can do so with https://github.com/quarto-ext/fontawesome which embed the fonts in the extension directly. 

The issue you mentioned with Google fonts is about Fonts for the text or the source code sans serif font and monospace fonts. 

You can use whatever you want from wherever you want. Usually I am using google fonts, so I don't know about other services. Though you can also easily use local fonts if you download them

**Comment by chuxinyuan** [comment]
@cderv @mcanouil Thank you for taking the time to solve this problem. I did the following:

## Add fonts.scss to the theme parameter

```yaml
format:
  html:
    theme: [cosmo, fonts.scss]
```

## fonts.scss code

```scss
/*-- scss:defaults --*/
$font-family-sans-serif: "Source Sans Pro", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "source-han-serif-sc", "Source Han Serif SC", "Source Han Serif CN", "Source Han Serif TC", "Source Han Serif TW", "Source Han Serif", "Noto Serif CJK SC", "Songti SC", "Microsoft YaHei", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol" !default; 

$font-family-monospace: "Lucida Console", 'Andale Mono', 'STKaiti', 'KaiTi', 'SimKai', monospace;
```

I don't know if the order of the fonts is correct, which temporarily solved my problem.

**Reply by mcanouil** [reply]
The order defines the fallback fonts when not found.
If "Source Sans Pro" and "Lucida Console" are found, all the others are not used.

**Reply by cderv** [reply]
I am surprised just setting this solved your issue, because it does not change where the font is retrieved, and it does look like the same as what is set by default 

Though, glad you solved your issue
